portal/api-authentication-grpc.go

```diff
@@ -0,0 +1,211 @@
+// and returning a hashed token.
+func (s *server) Login(ctx context.Context, in *pb.LoginRequest) (*pb.LoginResponse, error) {
+	log.Printf("Calling Login")
+	// Create Credentials
+	// TODO: validate credentials: username->email, tenant->shortname?
+	var res pb.LoginResponse
+	tenantName := in.GetCompany()
+	email := in.GetEmail()
+	pwd := in.GetPassword()
+
+	// Search for the tenant on the database
+	tenant, err := getTenant(tenantName)
+	if err != nil {
+		err = errors.New("Tenant not found")
+		res.Error = err.Error()
+		return &res, err
+	}
+
+	// Password validation
+	user, ok := getUser(tenant.Name, email)
+	// If a password exists for the given user
+	// AND, if it is the same as the password we received, then we can move ahead
+	expectedPwd := user.Password
+	// TODO: password will come not hashed and stored hashed so we need to hash it and compare it against db
+	if !ok || expectedPwd != pwd {
+		err := errors.New("Wrong tenant, email or password")
+		res.Error = err.Error()
+		return &res, err
+	}
+
+	// Add the session within a transaction in case anything goes wrong during the adding process
+	db := cluster.GetInstance().Db
+	tx, err := db.BeginTx(ctx, nil)
+	if err != nil {
+		tx.Rollback()
+		res.Error = err.Error()
+		return &res, err
+	}
+
+	// Set query parameters
+	loginCtx := cluster.NewContext(ctx, tx)
+	// Insert a new session with random string as id
+	sessionID, err := GetRandString(32, "sha256")
+	if err != nil {
+		tx.Rollback()
+		res.Error = err.Error()
+		return &res, err
+	}
+	userID := uuid.NewV4()
+
+	query :=
+		`INSERT INTO
+				m3.provisioning.sessions ("id","user_id", "tenant_id", "occurred_at")
+			  VALUES
+				($1,$2,$3,$4)`
+
+	// Execute Query
+	_, err = loginCtx.Tx.Exec(query, sessionID, userID, tenant.ID, time.Now())
+	if err != nil {
+		tx.Rollback()
+		res.Error = err.Error()
+		return &res, err
+	}
+	fmt.Println("sessionID: ", sessionID)
+
+	// Return session in Token Response
+	res.JwtToken = sessionID
+
+	// if no error happened to this point commit transaction
+	err = tx.Commit()
+	if err != nil {
+		res.Error = err.Error()
+		return &res, err
+	}
+	return &res, nil
+}
+
+// getUser searches for the user in the defined tenant's database
+// and returns the User if it was found
+func getUser(tenant string, email string) (user User, ok bool) {
+
+	bgCtx := context.Background()
+	db := cluster.GetInstance().GetTenantDB(tenant)
+
+	tx, err := db.BeginTx(bgCtx, nil)
+	if err != nil {
+		tx.Rollback()
+		return user, false
+	}
+	loginCtx := cluster.NewContext(bgCtx, tx)
+
+	// Get user from tenants database
+	var userEmail string
+	quoted := pq.QuoteIdentifier(tenant)
+	queryUser := fmt.Sprintf(`
+		SELECT 
+				t1.id, t1.email, t1.password
+			FROM 
+				tenants.%s.users t1
+			WHERE email=$1`, quoted)
+	row := loginCtx.Tx.QueryRow(queryUser, email)
+
+	// Save the resulted query on the User struct
+	err = row.Scan(&user.UUID, &userEmail, &user.Password)
```

Comment:
I cannot find where this `user` variable is instanciated

Reply:
it is being created on the definition of the function 
`func getUser(tenant string, email string) (user User, ok bool)`